horovod/common/operations.cc

```diff
@@ -954,6 +1126,61 @@ void BackgroundThreadLoop(HorovodGlobalState& state, MPIContext& ctx) {
   }
 }
 
+// If all messages in queue have responses in cache, use fast path with
+// no additional MPI coordination.
+void RunBypass(std::queue<Request>& message_queue, CacheCoordinator& cache_coordinator,
+               HorovodGlobalState& state, MPIContext& ctx) {
+
+  // Convert cache hits to responses. Populate so that least
+  // recently used responses get priority.
+  std::deque<Response> responses;
+  for (auto bit : cache_coordinator.cache_hits()) {
+    responses.push_back(state.response_cache.get_response(bit));
+  }
+
+  // Fuse responses as normal.
+  auto response_list = FuseResponses(responses, state, ctx);
+
+  if (!response_list.responses().empty()) {
```

Comment:
Seems like lines 1141-1181 are currently copy-pasted three times, can we extract them as a helper functions?

Reply:
Alright, I tried to consolidate some of the repeated code. I wrote a small helper function to get the tensor names and size data for the autotuner. I also merged the coordinator and worker rank logic after the response list broadcast to finally remove some code repetition that has existed in the code for a while. The remaining repeated code is just a few function calls now and logging code.

Reply:
Maybe we want to merge it as well, but perhaps another time :-)